### PR TITLE
De-convolution layer

### PR DESCRIPTION
**Pull request by pluskid**
This is a working in process PR to address #46. Basically this add an option to the `ConvolutionLayer` to allow it perform _de-convolution_ instead. A _de-convolution_ is conceptually like a "inverted" convolution layer, or could also be thought as a convolution layer with fractional stride. See http://arxiv.org/abs/1411.4038 for more details. The Caffe model definition for the mentioned paper could be found [here](https://gist.github.com/longjon/d24098e083bec05e456e#file-readme-md), including pointer to an implementation of deconvolution in Caffe (not merged into the official Caffe repo yet).

### CODE REVIEW

**Comment by philtomson** [comment]
@pluskid Thanks for adding this. We're wondering if there might be an error in backpropagation for the deconvolution layer (ConvolutionLayer with deconv=true). We keep seeing the Square loss go very rapidly towards Inf or NaN.

Network looks like this:

      NAME: ImageTrain
   BACKEND: CPUBackend

ARCHITECTURE: 5 layers
............................................................
*** MemoryDataLayer(data)
Outputs ---------------------------
data: Blob(750 x 750 x 3 x 10)
............................................................
*** MemoryDataLayer(label)
Outputs ---------------------------
label: Blob(750 x 750 x 10)
............................................................
*** ConvolutionLayer(conv1)
Inputs ----------------------------
data: Blob(750 x 750 x 3 x 10)
Outputs ---------------------------
conv: Blob(746 x 746 x 16 x 10)
............................................................
*** ConvolutionLayer(deconv)
Inputs ----------------------------
conv: Blob(746 x 746 x 16 x 10)
Outputs ---------------------------
deconv: Blob(750 x 750 x 1 x 10)
............................................................
*** SquareLossLayer(loss)
Inputs ----------------------------
deconv: Blob(750 x 750 x 1 x 10)
label: Blob(750 x 750 x 10)

23-Jan 16:06:05:INFO:root:
23-Jan 16:06:05:INFO:root:## Performance on Validation Set after 0 iterations
23-Jan 16:06:05:INFO:root:---------------------------------------------------------
23-Jan 16:06:05:INFO:root: Square-loss (avg over 10) = 77664.4219
23-Jan 16:06:05:INFO:root:---------------------------------------------------------
23-Jan 16:06:05:INFO:root:
23-Jan 16:07:54:INFO:root:000002 :: TRAIN obj-val = 1906052777742086025773056.00000000
23-Jan 16:10:16:INFO:root:
23-Jan 16:10:16:INFO:root:## Performance on Validation Set after 4 iterations
23-Jan 16:10:16:INFO:root:---------------------------------------------------------
23-Jan 16:10:16:INFO:root: Square-loss (avg over 10) = Inf
23-Jan 16:10:16:INFO:root:---------------------------------------------------------
23-Jan 16:10:16:INFO:root:

**Comment by pluskid** [comment]
@philtomson First of all, could you please run the unit-test locally? I have only tested the code in my local Mac only. I have not tested on my Linux server node yet because I do not have very stable Internet connection during this month. But the Travis-CI shows test error with de-convolution. If you also run into unit-test error on de-convolution, please let me know, I will try to see if there is any bug in the code.

If the unit-test passed, the computation should be correct, at least consistent with the unit-test. The unit-test code is written in the most explicit form, thus should be easy to check its semantics. That being said, even if the computation is correct, this kind of blow up is still possible (even in normal deep neural networks). If I remember correctly, in the original paper of de-convolution, they use quite sophisticated pre-training and train the network stage-by-stage. Maybe as a first try, you could try to add regularizer or norm constraints to the filter parameters of convolution and deconvolution layers.

**Comment by philtomson** [comment]
Chiyuan,

I'm running on Ubuntu 12.04.

This is what I'm seeing when running the unit tests:
...passing tests...
-- Testing Convolution(frozen=false) on CPUBackend{Float32} filter=(3,4)...
    > Setup
    > Forward
    > Backward
-- Testing De-Convolution(frozen=false) on CPUBackend{Float32}
filter=(3,4)...
    > Setup
    > Forward
    > Backward
ERROR: test failed: is_grad_bias_match
 in error at error.jl:21
 in default_handler at test.jl:19
 in do_test at test.jl:39
 in test_convolution_layer at
/home/phil/build/Mocha.jl/test/layers/convolution.jl:82
 in test_convolution_layer at
/home/phil/build/Mocha.jl/test/layers/convolution.jl:239
 in test_convolution_layer at
/home/phil/build/Mocha.jl/test/layers/convolution.jl:243
 in test_convolution_layer at
/home/phil/build/Mocha.jl/test/layers/convolution.jl:253
 in include at ./boot.jl:245
 in include_from_node1 at ./loading.jl:128
 in anonymous at /home/phil/build/Mocha.jl/test/runtests.jl:35
 in map at abstractarray.jl:1328
 in test_dir at /home/phil/build/Mocha.jl/test/runtests.jl:34
 in include at ./boot.jl:245
 in include_from_node1 at loading.jl:128
 in process_options at ./client.jl:285
 in _start at ./client.jl:354
while loading /home/phil/build/Mocha.jl/test/layers/convolution.jl, in
expression starting on line 256
while loading /home/phil/build/Mocha.jl/test/runtests.jl, in expression
starting on line 38

I was just doing some debugging in convolution.jl and noticed (comments in
*bold)*

Debug.@debug function forward(backend::CPUBackend,
state::ConvolutionLayerState, inputs::Vector{Blob})
  for i = 1:length(inputs)
    input = inputs[i]
    output = state.blobs[i]
    erase!(output)

    if state.layer.deconv
     * conv_bwd_impl(backend, state, input, output) #deconv is true so this
runs*
    else
      conv_fwd_impl(backend, state, input, output)
    end
Debug.@bp

*#at this point most entries in output.data were rather large*

*#negative numbers, see below*
    # add bias
    width_out, height_out, channels_out, num_out = size(output)
    top_img_offset = width_out*height_out*channels_out * sizeof(state.dtype)

    for n = 1:num_out
      output_ptr = convert(Ptr{state.dtype}, output.data) + top_img_offset
* (n-1)
      RawBLAS.gemm!('N', 'N', state.width_out*state.height_out,
state.layer.n_filter, 1, convert(state.dtype, 1),
          state.etc.bias_multiplier.data, pointer(state.bias.data),
convert(state.dtype, 1), output_ptr)
*#at this point all entries in output.data are -Inf*
    end
  end
end


Here's a sample of the values in output.data at the breakpoint:

0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 -2.4492218e6
-8.871836e6 -1.915679e7 -3.1606272e7 -4.6570972e7 -5.7393468e7 -6.207661e7
-6.104596e7 -5.5546204e7 -4.54581e7 .... (many more non-0 values, though)

I checked state.bias.data going into that last for-loop:

debug:300> state.bias.data
Float32[-2.883422e8,-6.4813747e9,-6.9414257e9,-2.8844653e8,-2.5666177,-3.2903498e9,-2.926552e9,-4.861353e8,-306.26486,-2.2729994e8,-3.971654e9,-8.135067e8,-8.338445e9,-22.690199,-88.82442,-5.2384886e8]

...so perhaps we do have some scaling issues somewhere as you suggest,
however, the unit test for deconvolution also fails.

Phil

**Comment by pluskid** [comment]
@philtomson Thanks for sharing the information. The very bad bias might come from ill conditioned gradient updates. However, as you discovered, there seem to be a bug in de-convolution when running under Linux. I will try to look into this issue soon.

BTW: Sometimes normalizing your input data to have zero mean and unit standard deviation helps if you have not tried that yet.

**Comment by pluskid** [comment]
@philtomson BTW: I forgot to mention that I have tested on my Linux node: the unit-test failure is due to floating point issues. By relaxing the test threshold a bit, I could have the test passed.

**Comment by vchuravy** [comment]
What is holding this up? I would be interested in using deconvolution for image segmentation and would be willing to help out with testing.

**Comment by pluskid** [comment]
@vchuravy This is holding up because the implementation is not complete (only CPU was implemented). And yes, there is already a GPU implementation in mxnet. So you could actually try out that with [MXNet.jl](https://github.com/dmlc/MXNet.jl). See [the doc here](http://mxnetjl.readthedocs.org/en/latest/api/symbol.html#Deconvolution).
